src/control/server/ctl_ranks_rpc.go

```diff
@@ -424,46 +422,45 @@ func (svc *ControlService) SetEngineLogMasks(ctx context.Context, req *ctlpb.Set
 		return nil, errors.New("nil request")
 	}
 
-	errs := []string{}
+	resp := new(ctlpb.SetLogMasksResp)
+	instances := svc.harness.Instances()
+	resp.Errors = make([]string, len(instances))
 
-	for idx, ei := range svc.harness.Instances() {
+	for idx, ei := range instances {
 		eReq := *req // local per-engine copy
 
+		if int(ei.Index()) != idx {
+			svc.log.Errorf("engine instance index %d doesn't match engine.Index %d",
+				idx, ei.Index())
+		}
+
 		if !ei.IsReady() {
-			errs = append(errs, fmt.Sprintf("engine-%d: not ready", ei.Index()))
+			resp.Errors[idx] = "not ready"
 			continue
 		}
 
 		if err := updateSetLogMasksReq(svc.srvCfg.Engines[idx], &eReq); err != nil {
-			errs = append(errs, errors.Wrapf(err, "engine-%d", ei.Index()).Error())
+			resp.Errors[idx] = err.Error()
 			continue
 		}
 		svc.log.Debugf("setting engine %d log masks %q, streams %q and subsystems %q",
 			ei.Index(), eReq.Masks, eReq.Streams, eReq.Subsystems)
 
 		dresp, err := ei.CallDrpc(ctx, drpc.MethodSetLogMasks, &eReq)
 		if err != nil {
-			errs = append(errs, fmt.Sprintf("engine-%d: %s", ei.Index(), err))
+			resp.Errors[idx] = err.Error()
 			continue
 		}
 
 		engineResp := new(ctlpb.SetLogMasksResp)
 		if err = proto.Unmarshal(dresp.Body, engineResp); err != nil {
-			errs = append(errs, fmt.Sprintf("engine-%d: %s", ei.Index(), err))
-			continue
+			return nil, err
```

Comment:
Why do you not set this error in the response (as for the other error) ?

Reply:
The other errors could be a result of expected failures related to the responding server being unavailable where we want to continue to get a response from one engine if the other is not contactable. This error indicates a serious problem with protocol incompatibility or a programming error which should be immediately fatal.

Reply:
OK, thanks for the answer.